synapse/handlers/federation.py

```diff
@@ -279,8 +279,11 @@ async def _maybe_backfill_inner(
         # events.
         #
         # We do this by filtering all the backwards extremities and seeing if
-        # any remain. Given we don't have the extremity events themselves, we
-        # need to actually check the events that reference them.
+        # any remain.
```

Comment:
Assuming what I've written above is vaguely sensible, I'd suggest a nested function like:

```python

    def visible_successors(backfill_point: _BackfillPoint) -> List[str]:
        ...
        # compute filtered_extremities as before, then return it
```

or perhaps a similar free function taking a `store` and `event_id`. Then 

```python
if not found_filtered_extremity and not insertion_events_to_be_backfilled:
```

could become e.g.

```python
if (
    not any(visible_successor(bp) for bp in backwards_extremities)
    and not insertion_events_to_be_backfilled
):
    ...
```

(I wouldn't normally make as much fuss on arguable readability/style things, but we're in the bowels of a 200 line function here)

Reply:
1f0a63e6b57ca94775408c3b3ad81ff0df3dbceb changes the situation somewhat. Let's take a look at the whole when I get to the last commit...

Reply:
In case it helps, this is talking about "filtering" in the sense of "a call to `filter_events_for_server`".

(and yes, that's really not clear when first reading.)